feat(PPDSC-1982): layout props 2 of 6

Xin00163 opened the pull request:
PPDSC-1982

1. Add logical margin and padding support to the following components and provide storybook examples:

> Aspect Ratio (Do not need logical properties)
> Audio Player
> Banner
> Byline
> Caption (Caption will be rebuilt in https://nidigitalsolutions.jira.com/browse/PPDSC-2091)
> Card
> Checkbox (Radio Button is done as well)
> Date Time

2. Provide storybook examples for components in https://github.com/newscorp-ghfb/newskit/pull/84:
> Block
> TextBlock
> Stack
> Button
> Tabs
> AssistiveText
> Label
> Link
Select will be in a separate ticket.
3. Apply logical props to the sub-components if it has nested structure like the card.
4. Expand functions in logical-properties

There should be no visual changes to the components and no breaking changes.

<!---
This section will be used to indicate if we should move to a major version in the next release, remove if not revelant.
DO CONSIDER any of the following are breaking changes to a consumer, this is by no mean an exhaustive list.
-removing or renaming props
-removing or renaming tokens
-removing or renaming components
-removing or renaming exported functions
-in some cases, major bumps to peer dependencies
--->


**I have done:**
- [x] Written unit tests against changes
- [ ] Written functional tests against the component and/or NewsKit site
- [ ] Updated relevant documentation

**I have tested manually:**
- [ ] The feature's functionality is working as expected on Chrome, Firefox, Safari and Edge
- [ ] The screen reader reads and flows through the elements as expected.
- [ ] There are no new errors in the browser console coming from this PR.
- [ ] When visual test is not added, it renders correctly on different browsers and mobile viewports (Safari, Firefox, small mobile viewport, tablet)
- [ ] The Playground feature is working as expected


<!---
Below sections are optional
--->

**Before:**
<!--- Drag and Drop your screenshot's here --->

**After:**
<!--- Drag and Drop your screenshot's here --->

**Who should review this PR:**
<!---
If you know someone is a domain expert for your PR,
someone who is deeply involved in the story,
ask them explicitly to review the PR.
--->

**How to test:**
<!--
If it's not immediately obvious how to test this PR, give instructions.
It's mandatory to update README.MD or development documentation if existing test strategy had changed.
-->

<!--
More info about raising an good PR: https://nidigitalsolutions.jira.com/wiki/spaces/NPP/pages/1319370846/Pull+Request
-->

newskit-engineering commented:
You can preview these changes on: 
  Documentation Site: http://ncu-newskit-docs.s3-website-eu-west-1.amazonaws.com/ppdsc-1982-layout-props-2-of-6 
  Storybook: http://ncu-newskit-docs.s3-website-eu-west-1.amazonaws.com/ppdsc-1982-layout-props-2-of-6/storybook

mutebg commented on src/utils/logical-properties.ts:
```diff
@@ -22,90 +22,141 @@ interface LogicalPadding {
-  overrides: Record<string, unknown>,
+  // https://github.com/microsoft/TypeScript/issues/15300
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  overrides: any,
```
as I mentioned before you can try making this `object  | undefined`

mutebg commented on src/banner/styled.tsx:
```diff
@@ -31,27 +36,43 @@ export const StyledIconContentContainer = styled(Stack)<
       'spaceInline',
-    )}
+    )};
+  ${({layout}) =>
+    layout === 'vertical'
+      ? logicalPropsWithCSSProperty(
+          'marginBlockEnd',
+          `banner.${layout}.content`,
+          'content',
+        )
+      : logicalPropsWithCSSProperty(
+          'marginInlineEnd',
+          `banner.${layout}.content`,
+          'content',
+        )};
```
I believe `spaceInline` prop is actually the `gap` between the icon and container here. so that props eventually need renaming. and eventually the code should look something like
```js
${({layout}) =>
    getResponsiveSpace(
      layout === 'vertical' ? 'marginBlockEnd' : 'marginInlineEnd',
      `banner.${layout}.content`,
      'content',
      'gap',
    )};
```

Xin00163 replied:
Have removed the change and left a comment in the types.ts file

Xin00163 commented on src/card/types.ts:
```diff
@@ -12,12 +13,22 @@ export interface CardOverridesProps {
+     * @deprecated This property is deprecated and will be removed in the next major release. Use `PaddingInline` & `PaddingBlock` instead.
+     */
     spaceInset?: MQ<string>;
+    paddingBlock?: MQ<string>;
+    paddingInline?: MQ<string>;
```
They are used to replace `spaceInset`

Xin00163 commented on src/base-switch/styled.tsx:
```diff
@@ -47,14 +53,26 @@ export const StyledSwitchContainer = styled.div<
       'input',
       'spaceInline',
     )}
+    ${({size, labelPosition, path}) =>
+    labelPosition === 'end'
+      ? logicalPropsWithCSSProperty(
+          'marginInlineEnd',
+          `${path}.${size}.input`,
+          'input',
+        )
+      : logicalPropsWithCSSProperty(
+          'marginInlineStart',
+          `${path}.${size}.input`,
+          'input',
+        )}
```
Does it mean we should no try to use logical properties in this case?

Xin00163 commented:
Should have addressed all the comments @mutebg. 
Some of the `spaceInline` & `spaceStack` properties are referring to the gap between two elements. As discussed with @mutebg, comments have been added to highlight those cases. When we are ready for a breaking change ticket, we could filter those comments and change the `spaceInline` & `spaceStack` property name to `gap` or other names.

I noticed that there are no unit tests for logical props added to first batch of components. I didn't add any either as I think logical-properties.test.ts should cover it. Let me know if you think they are needed.

mutebg commented on src/card/types.ts:
```diff
@@ -12,12 +13,30 @@ export interface CardOverridesProps {
+     * @deprecated This property is deprecated and will be removed in the next major release. Use padding logical props instead.
+     */
     spaceInset?: MQ<string>;
+    paddingBlock?: MQ<string>;
+    paddingBlockStart?: MQ<string>;
+    paddingBlockEnd?: MQ<string>;
+    paddingInline?: MQ<string>;
+    paddingInlineStart?: MQ<string>;
+    paddingInlineEnd?: MQ<string>;
```
Can we just use `LogicalPadding` here ?

Xin00163 replied:
OK. Will need to export this type

Vanals commented on site/pages/components/aspect-ratio.mdx:
```diff
@@ -6,7 +6,7 @@ import {AspectRatio, getAspectRatioStyles} from 'newskit';
 export default Layout;
 
-# Aspect Ratio
+# (deprecated) Aspect Ratio
```
Why are we deprecating it? Should we also explain it to the user/propose the alternative?

Xin00163 replied:
It was decided in yesterday's tech analysis as this component is not used. Maybe I should point the user to the `getAspectRatioStyles` util function used in images?

Vanals replied:
Yes I think we should 👍

Xin00163 replied:
I think this should be done in a different ticket if we think there is a need. `getAspectRatioStyles` is an internal util function at the moment. The thing is we do not think there is a use case for users to use `AspectRatio` or `getAspectRatioStyles` on its own, hence it is deprecated. Unless you know there is use case?

For now, I will say "Please refer to the Image component's documentation about the loadingAspectRatio`. Any suggestions are welcome

Vanals replied:
Make sense, thank you Xin

Vanals commented on src/aspect-ratio/types.ts:
```diff
@@ -19,6 +19,9 @@ export interface StyledContainerProps {
 }
 
+/**
+ * @deprecated AspectRatioProps is deprecated and will be removed in the next major release.
```
Do we need a ticket for removing it? 🤔

Xin00163 replied:
It was decided that I should do it in this ticket.

pdimova replied:
we can mention the  Image's `loadingAspectRatio` prop

Vanals commented on src/banner/types.ts:
```diff
@@ -22,14 +26,14 @@ export interface BannerBaseProps extends React.AriaAttributes {
-      spaceInline?: MQ<string>;
+      spaceInline?: MQ<string>; // Note: spaceInline is used as a gap. Should rename this to gap instead of using logical props.
     };
     content?: {
-      spaceInline?: MQ<string>;
+      spaceInline?: MQ<string>; // Note: LogicalProps can not replace spaceInline! spaceInline = Stack's gap
       title?: {
         stylePreset?: MQ<string>;
         typographyPreset?: MQ<string>;
-        spaceStack?: MQ<string>;
+        spaceStack?: MQ<string>; // Note: spaceStack is used as a gap. Should rename this to gap instead of using logical props.
```
is this going to happen in future? when ready for the breaking change?

Xin00163 replied:
Yes, we will do this in a separate ticket when we are ready for a breaking change.

Vanals commented on src/byline/byline.tsx:
```diff
@@ -101,7 +105,12 @@ const ThemelessByline: React.FC<BylineProps> = ({bylineData, overrides}) => {
 
   const PipeContainer = styled.span<{overrides: BylineProps['overrides']}>`
-    margin-left: ${getSpacingFromTheme(dividerSpace)};
+    ${getResponsiveSpace(
+      'marginLeft',
+      'byline.divider',
+      'divider',
+      'spaceInline',
+    )}
```
🧠

Vanals commented on src/assistive-text/__tests__/assistive-text.stories.tsx:
```diff
@@ -128,3 +132,62 @@ export const AssistiveTextEnhancers = () => (
+      </AssistiveText>
+    </StyledDiv>
+    <StorybookSubHeading>
+      Assistive Text With Logical Paddings and enhancers
```
Does having both padding and margin make sense for the assistive text?
I wonder if we should have only margin, given the `assistive text` it self has no border. 🤔

Vanals replied:
Shuldn't the padding apply only the the user can add a spacing withing the component it self? for example a button with wants to increase the space between the button text and the button container?
Same question for Audio player

Vanals replied:
What you did in the banner is what I would think as a good application of both padding and margin

Xin00163 replied:
I see what you mean and tbh I kind of agree. For example, checkbox with padding is kind of a weird situation. I had the discussion with @jps . The idea is to keep the logicalProps consistent across all the components. We can't have only margin no padding for one but both margin & padding for the other. At the end of the day, it is up to the user to decide what they want to add to the components and they should consider whether to add padding or margin. 

We can't predict the behaviour of our users, so we probably need to provide the means for them to do so.

Vanals replied:
I see, make sense I guess... . Also natively all the elements can take both padding and margin..then is up to the devs to use them wisely! Maybe was just weird for me to see them applied on storybook. But at least we test them.

Xin00163 replied:
I am hoping that @pdimova will have time to review this pr tomorrow, so won't merge it yet.

pdimova replied:
let's have a quick catch up when you both are online

pdimova replied:
On second thought, I have mixed feelings about the padding props for this particular component? 

Why do we need padding in general?

> Padding is used to create space around an element's content, inside of any defined borders.
https://www.w3schools.com/css/css_padding.asp

Which is the content inside the `WithEnhancers` container that we'd like to separate from each other? 
I don't see such, though

My understanding of the concept of padding makes me a bit reluctant to add the logical props here.
Let's further discuss it in a huddle

jps commented:
Crashing the party very late on this, but don't we need to update the API tables in the docs?

Xin00163 commented:
> Crashing the party very late on this, but don't we need to update the API tables in the docs?

Indeed @jps . Totally forgot about that.

pdimova commented on src/utils/logical-properties.ts:
```diff
@@ -36,88 +37,111 @@ const getResponsiveSpace = (
+  'paddingBlock',
+];
+
+export const omitLogicalPropsFromOverrides = (overrides: object | undefined) =>
```
what is the difference between `omitLogicalPropsFromOverrides` and `extractLogicalPropsFromOverrides`?

Xin00163 replied:
`omitLogicalPropsFromOverrides` will reject the logical props from overrides, while `extractLogicalPropsFromOverrides` will filter out the logical props.

pdimova commented on src/utils/logical-properties.ts:
```diff
@@ -36,88 +37,111 @@ const getResponsiveSpace = (
+export const omitLogicalPropsFromOverrides = (overrides: object | undefined) =>
+  rejectObject(overrides || {}, logicalPropsArray as keyof object);
+
+export const extractLogicalPropsFromOverrides = (
```
Couldn't find any tests for this util function :)

pdimova commented on src/utils/__tests__/logical-properties.test.ts:
```diff
@@ -66,19 +85,86 @@ describe('logical properties', () => {
     });
   });
+  it('should return an empty object if overrides is undefined', () => {
+    expect(extractLogicalPropsFromOverrides(undefined)).toEqual({});
```
not sure about this, why do we need an empty object out of an undefined value?

Xin00163 replied:
Because of this `filterObject(overrides || {}, logicalPropsArray as keyof object)`. If it's undefined, it will return `{}`. Otherwise, we will need to check if it's undefined each time we use it.

pdimova commented on src/assistive-text/__tests__/assistive-text.stories.tsx:
```diff
@@ -128,3 +132,62 @@ export const AssistiveTextEnhancers = () => (
+      <AssistiveText
+        startEnhancer={enhancer}
+        endEnhancer={enhancer}
+        overrides={{marginBlock: 'space050'}}
```
You can add one more story - both with padding and margin

Xin00163 replied:
Let's have a chat about this @pdimova

pdimova commented on src/assistive-text/assistive-text.tsx:
```diff
@@ -15,31 +16,35 @@ const ThemelessAssistiveText = ({
+        <StyledAssistiveText
+          aria-disabled={state === 'disabled' ? true : undefined}
+          size={size}
+          overrides={nonLogicalOverrides}
```
Basically, there is no need to filter out the logical props here since the  `StyledAssistiveText` internal implementation has no logic to apply them. 
But semantically it is good to highlight there props don't belong there

We have plenty of places where we pass unnecessary props inside the overrides object 

Here it is a good place to initiate a props clean-up  :)

pdimova commented on src/card/types.ts:
```diff
@@ -1,9 +1,10 @@
+export interface CardOverridesProps extends LogicalProps {
   stylePreset?: MQ<string>;
   horizontalRatio?: string;
   mediaContainer?: {
```
what about the spaceInline here?

pdimova commented on src/card/styled.tsx:
```diff
@@ -111,12 +113,14 @@ export const StyledCardContainerTeaser = styled.div<
     'teaserContainer',
     'spaceInset',
   )}
+  ${logicalProps('card.teaserContainer', 'teaserContainer')};
```
do we need margin here?

Xin00163 replied:
I only added padding because of `spaceInset` but I tried add margin and it seems OK. I could extend to accept both.

pdimova commented on src/card/styled.tsx:
```diff
@@ -153,12 +157,14 @@ export const StyledCardContainerActions = styled(Stack)<
     'actionsContainer',
     'spaceInset',
   )}
+  ${logicalProps('card.actionsContainer', 'actionsContainer')};
```
same as above? do we need margin? padding here is used only for the inset card

pdimova commented on src/card/card.tsx:
```diff
@@ -190,21 +194,21 @@ const ThemelessCard = React.forwardRef<HTMLDivElement, CardProps>(
               hasHref={hasHref}
               layout={layout}
-              overrides={overrides}
+              overrides={nonLogicalOverrides}
```
why passing nonLogicalOverrides here? I'm a bit confused why we need this

Xin00163 replied:
Good point. I will remove the cases where we don't need filter

pdimova commented on src/caption/types.ts:
```diff
@@ -1,17 +1,24 @@
+    /**
+     * @deprecated This property is deprecated and will be removed in the next major release.
+     */
+    spaceStack?: MQ<string>;
```
This is the gap, can not be deprecated

Xin00163 replied:
Good catch

pdimova commented on src/radio-button/__tests__/radio-button.stories.tsx:
```diff
@@ -346,3 +346,38 @@ export const StoryRadioButtonOverrides = () => {
+            value="1"
+            label="Option 1"
+            overrides={{
+              marginBlockEnd: 'space030',
```
The wrapper GridLayout sets a gap of 16px between the. buttons `rowGap="space040"` Maybe we can remove the Grid for this particular story

pdimova commented on src/card/styled.tsx:
```diff
@@ -59,21 +61,23 @@ export const StyledCardContainerMedia = styled.div<
+  // getResponsiveSpace to be removed once logical props are used in defaults
   ${handleResponsiveProp({layout: DEFAULT_PROPS.layout}, ({layout}, props) => {
     let marginDirection = 'marginBottom';
     if (isHorizontal(layout) && isReverse(layout)) {
       marginDirection = 'marginLeft';
     } else if (isHorizontal(layout)) {
       marginDirection = 'marginRight';
```
Is this logic still neded?

Xin00163 replied:
I think so until `spaceInline` is removed from defaults. I've add a comment to remove it later.

pdimova reviewed (approved):
Good job!
